plugin/http/propagation/b3/b3.go

```diff
@@ -0,0 +1,115 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package b3 contains a propagation.HTTPFormat implementation
+// for B3 propagation. See https://github.com/openzipkin/b3-propagation
+// for more details.
+package b3
+
+import (
+	"encoding/hex"
+	"net/http"
+
+	"go.opencensus.io/trace"
+	"go.opencensus.io/trace/propagation"
+)
+
+const (
+	traceIDHeader = "X─B3─TraceId"
+	spanIDHeader  = "X─B3─SpanId"
+	sampledHeader = "X─B3─Sampled"
+)
+
+// HTTPFormat implements propagation.HTTPFormat to propagate
+// traces in HTTP headers in B3 propagation format.
+type HTTPFormat struct{}
+
+var _ propagation.HTTPFormat = (*HTTPFormat)(nil)
+
+// FromRequest extracts a B3 span context from incoming requests.
+func (f *HTTPFormat) FromRequest(req *http.Request) (sc trace.SpanContext, ok bool) {
+	tid, ok := parseTraceID(req.Header.Get(traceIDHeader))
+	if !ok {
+		return trace.SpanContext{}, false
+	}
+	sid, ok := parseSpanID(req.Header.Get(spanIDHeader))
+	if !ok {
+		return trace.SpanContext{}, false
+	}
+	sampled, _ := parseSampled(req.Header.Get(sampledHeader))
+	return trace.SpanContext{
+		TraceID:      tid,
+		SpanID:       sid,
+		TraceOptions: sampled,
+	}, true
+}
+
+func parseTraceID(tid string) (trace.TraceID, bool) {
+	if tid == "" {
+		return trace.TraceID{}, false
+	}
+	b, err := hex.DecodeString(tid)
+	if err != nil {
+		return trace.TraceID{}, false
+	}
+	if len(b) < 8 { // TraceID needs to be at least 32-bits.
+		return trace.TraceID{}, false
+	}
+	var traceID trace.TraceID
+	if len(b) == 16 {
+		// 64-bit trace ID.
+		copy(traceID[:], b)
+	} else {
+		// 32-bit trace ID.
+		copy(traceID[8:], b[:8])
+	}
+	return traceID, true
+}
+
+func parseSpanID(sid string) (trace.SpanID, bool) {
+	if sid == "" {
+		return trace.SpanID{}, false
+	}
+	b, err := hex.DecodeString(sid)
+	if err != nil {
+		return trace.SpanID{}, false
+	}
+	if len(b) != 8 {
+		return trace.SpanID{}, false
+	}
+	var spanID trace.SpanID
+	copy(spanID[:], b)
+	return spanID, true
+}
+
+func parseSampled(sampled string) (trace.TraceOptions, bool) {
+	switch sampled {
+	case "0":
+		return trace.TraceOptions(0), true
+	case "1":
+		return trace.TraceOptions(1), true
+	default:
+		return trace.TraceOptions(0), false
+	}
+}
+
+// ToRequest modifies the given request to include B3 headers.
+func (f *HTTPFormat) ToRequest(sc trace.SpanContext, req *http.Request) {
+	req.Header.Set(traceIDHeader, hex.EncodeToString(sc.TraceID[:]))
+	req.Header.Set(spanIDHeader, hex.EncodeToString(sc.SpanID[:]))
+
+	if sc.IsSampled() {
```

Comment:
b3 is tri-state. so if IsSampled false is an explicit decision to not sample, it should be propagated as zero. Otherwise the next hop will re-sample, which is uncommon in B3 as they trust upstream by default

Reply:
This is related to the issue above. For OpenCensus, zero means there is no sampling decision and the next hop might need to determine whether they should sample or not. Let's try to address https://github.com/census-instrumentation/opencensus-go/pull/380/files#r164938998 first.